### PR TITLE
the working structjump code

### PR DESCRIPTION
**Pull request by fqiang**
+ added the othermap instead of othervar. used to track of the mapping of parent -> child dummy variable indices. 
+ added function for computing list of scenario IDs on a local process. 
+ added some StructJuMP example models that can be solved by current supported solvers, Ipopt and PIPS.

### CODE REVIEW

**Comment by mlubin** [comment]
@joehuchette, now that PIPS is publicly released, could we set up travis CI here?

File: src/nlp.jl [hunk 1]
```diff
@@ -1,12 +1,26 @@
 # Add parent::Int here to avoid overwritting same method in JuMP. Should be safe.
 function JuMP.parseNLExpr_runtime(m::JuMP.Model, x::JuMP.Variable, tape, parent::Int, values)
+    # @show m
+    # @show parent
+    # @show JuMP.getName(x)
+    # @show x.col
+    # @show tape
+    # @show parent
+    # @show values
```
Comment (joehuchette) [inline src/nlp.jl]: Can we remove these commented-out lines?

File: src/nlp.jl [hunk 2]
```diff
@@ -1,12 +1,26 @@
...
-        others = getStructure(m).othervars
-        push!(others, x)
-        push!(tape, JuMP.NodeData(ReverseDiffSparse.EXTRA, length(others), parent))
+        othermap = getStochastic(m).othermap
+        if haskey(othermap, x)
+            newx = othermap[x]
+        else
+            newx = JuMP.Variable(m,JuMP.getLower(x),JuMP.getUpper(x),:Cont, JuMP.getName(x))
+            othermap[x] = newx  #parent -> dummy parent in the child node
+            @show x.col, "-->",newx.col
```
Comment (joehuchette) [inline src/nlp.jl]: remove?

File: src/nlp.jl (hunk 1 above)
Comment (fqiang) [inline src/nlp.jl]: @joehuchette sure. delete them. Need to create another pull for this? (not very clear how github works yet).

**Comment by joehuchette** [comment]
Since it's PIPS specific, it seems like ``src/pips_structure_interface.jl`` should go in [PIPS.jl](https://github.com/kibaekkim/PIPS.jl) or somewhere similar?

File: src/nlp.jl (hunk 1 above)
Comment (joehuchette) [inline src/nlp.jl]: You can just remove them in another commit and push it to your local fork

**Comment by fqiang** [comment]
@joehuchette Julia Structure interface for PIPS-NLP is at repository in [PIPS](https://github.com/Argonne-National-Laboratory/PIPS/blob/master/PIPS-NLP/JuliaInterface/ParPipsNlp.jl), which I think serves the same purpose as the [PIPS.jl](https://github.com/kibaekkim/PIPS.jl). 
pips_structure_interface.jl is a glue code between them. I think it can go with either one. However I think it is a good demonstration of the usage of the StructJuMP.

**Comment by fqiang** [comment]
@joehuchette maybe put it in another folder, but I am not sure what's the best way to present this.

**Comment by joehuchette** [comment]
I really think the PIPS specific code needs to go somewhere else; this repo should only include solver-specific modeling code. PIPS.jl seems like the natural place to put it.

File: examples/pips/select_solver.jl [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    using SerialIpoptInterface
+    structJuMPSolve = SerialIpoptInterface.structJuMPSolve
+elseif ARGS[1] == "ParPips"
+    using ParPipsNlpInterface
+    structJuMPSolve = ParPipsNlpInterface.structJuMPSolve
+elseif ARGS[1] == "Pips"
+    using SerialPipsNlpInterface
+    structJuMPSolve = SerialPipsNlpInterface.structJuMPSolve
+else
+    Base.warn("Known solvers are: Ipopt, ParPips, or Pips")
```
Comment (joehuchette) [inline examples/pips/select_solver.jl]: Should probably be ``error``

File: src/StructJuMP.jl [hunk 4]
```diff
@@ -50,6 +52,27 @@ getchildren(m::JuMP.Model)    = getStructure(m).children
 getprobability(m::JuMP.Model) = getStructure(m).probability
 num_scenarios(m::JuMP.Model)  = getStructure(m).num_scen
 
+
+function getMyRank()
+    myrank = 0;
+    mysize = 1;
+    if isdefined(:MPI)==true && MPI.Initialized()==true && MPI.Finalized() == false
```
Comment (joehuchette) [inline src/StructJuMP.jl]: Can just be ``if isdefined(:MPI) && MPI.Initialized() && !MPI.Finalized()``

File: src/StructJuMP.jl [hunk 5]
```diff
@@ -50,6 +52,27 @@ getchildren(m::JuMP.Model)    = getStructure(m).children
...
+        comm = MPI.COMM_WORLD
+        mysize = MPI.Comm_size(comm)
+        myrank = MPI.Comm_rank(comm)
+    end
+    return myrank,mysize
+end
+
+function getScenarioIds(m::JuMP.Model)
```
Comment (joehuchette) [inline src/StructJuMP.jl]: Why does this function always return ``0``?

File: src/nlp.jl [hunk 6]
```diff
@@ -4,9 +4,15 @@ function JuMP.parseNLExpr_runtime(m::JuMP.Model, x::JuMP.Variable, tape, parent:
...
-        others = getStructure(m).othervars
-        push!(others, x)
-        push!(tape, JuMP.NodeData(ReverseDiffSparse.EXTRA, length(others), parent))
+        othermap = getStructure(m).othermap
+        if haskey(othermap, x)
+            newx = othermap[x]
+        else
+            newx = JuMP.Variable(m,JuMP.getlowerbound(x),JuMP.getupperbound(x),:Cont, JuMP.getname(x))
+            othermap[x] = newx  #parent -> dummy parent in the child node
+            #@show x.col, "-->",newx.col
```
Comment (joehuchette) [inline src/nlp.jl]: Remove comment?

**Comment by joehuchette** [comment]
Thanks @fqiang, looks good now. I have a few small comments. Can you please address those, and then squash this down to a single commit?

File: src/StructJuMP.jl (hunk 5 above)
Comment (fqiang) [inline src/StructJuMP.jl]: @joehuchette The root node (corresponding to id=0) is assumed to be allocated on every processes.

File: examples/pips/select_solver.jl (hunk 3 above)
Comment (fqiang) [inline examples/pips/select_solver.jl]: agree. modified to use Base.error

File: src/nlp.jl (hunk 6 above)
Comment (fqiang) [inline src/nlp.jl]: removed.

File: src/StructJuMP.jl (hunk 4 above)
Comment (fqiang) [inline src/StructJuMP.jl]: it can. updated the condition.

File: src/StructJuMP.jl (hunk 5 above)
Comment (joehuchette) [inline src/StructJuMP.jl]: Since this is assumed then, it might make sense to return just the child ranks

**Comment by fqiang** [comment]
@joehuchette Thanks. Just merge the commits into 35e7412. Let me know if that works.

**Comment by joehuchette** [comment]
Didn't seem to work, there are still 73 commits, and merge conflicts with master.

File: src/StructJuMP.jl (hunk 5 above)
Comment (fqiang) [inline src/StructJuMP.jl]: then, the solver will need to compute the set based on the the process rank. How about move this function to StructJuMPSolverInterface module?

**Comment by fqiang** [comment]
resolved the conflict, do you know the command to merge commits into one. or I will need to do another PR to achieve this? @joehuchette

File: src/StructJuMP.jl (hunk 5 above)
Comment (joehuchette) [inline src/StructJuMP.jl]: Sure

**Comment by joehuchette** [comment]
You can use ``git rebase -i master``. For each commit besides the first, change ``pick`` to ``f`` and step through the rebase. You can just open another PR if you run into issues, though.

**Comment by fqiang** [comment]
close this one to open a new PR.

**Pull request closed without merging**